File: JavaScript/Leetcode70.js
/**
 * @param {number} n
 * @return {number}
 */

// 爬楼梯
// 有多少种不同的方法可以爬到楼顶

// f(x)=f(x−1)+f(x−2)

// 递归-时间超限
var climbStairs = function (n) {
    if (n == 1) return 1
    if (n == 2) return 2
    return climbStairs(n - 1) + climbStairs(n - 2)
};

// 一个一个算 p+q算法^_^
var climbStairs = function (n) {
    let p = 0;
    let q = 0;
    let r = 1;

    for (let i = 1; i <= n; i++) {
        p = q;
        q = r;
        r = p + q;
    }
    return r;
};

File: Java/Leetcode70.java
public class Leetcode70 {

}

// 爬楼梯

// 动态规划

// 当前这一步依赖上一步的决策
// f(x)=f(x−1)+f(x−2)
class Solution {
    public int climbStairs(int n) {
        int p = 0, q = 0, r = 1;
        for (int i = 1; i <= n; i++) {
            p = q;
            q = r;
            r = p + q;
        }
        return r;
    }
}

// 矩阵快速幂
class Solution2 {
    public int climbStairs(int n) {
        int[][] q = { { 1, 1 }, { 1, 0 } };
        int[][] res = pow(q, n);
        return res[0][0];
    }

    public int[][] pow(int[][] a, int n) {
        int[][] ret = { { 1, 0 }, { 0, 1 } };
        while (n > 0) {
            if ((n & 1) == 1) {
                ret = multiply(ret, a);
            }
            n >>= 1;
            a = multiply(a, a);
        }
        return ret;
    }

    // 矩阵乘法
    public int[][] multiply(int[][] a, int[][] b) {
        int[][] c = new int[2][2];
        for (int i = 0; i < 2; i++) {
            for (int j = 0; j < 2; j++) {
                c[i][j] = a[i][0] * b[0][j] + a[i][1] * b[1][j];
            }
        }
        return c;
    }
}

// 通项公式

// 由递推公式得到通项公式
class Solution3 {
    public int climbStairs(int n) {
        double sqrt5 = Math.sqrt(5);
        double fibn = Math.pow((1 + sqrt5) / 2, n + 1) - Math.pow((1 - sqrt5) / 2, n + 1);
        return (int) Math.round(fibn / sqrt5);
    }
}